src/effects/backends/builtin/compressoreffect.cpp

```diff
@@ -0,0 +1,238 @@
+                        "Depending on the input signal, short release times "
+                        "may introduce a 'pumping' effect and/or distortion"));
+    release->setValueScaler(EffectManifestParameter::ValueScaler::Integral);
+    release->setUnitsHint(EffectManifestParameter::UnitsHint::Millisecond);
+    release->setRange(0, 150, 1500);
+
+    EffectManifestParameterPointer gain = pManifest->addParameter();
+    gain->setId("gain");
+    gain->setName(QObject::tr("Output gain"));
+    gain->setShortName(QObject::tr("Gain"));
+    gain->setDescription(
+            QObject::tr("The Output gain knob adjusts the level of the output "
+                        "signal after the compression was applied"));
+    gain->setValueScaler(EffectManifestParameter::ValueScaler::Linear);
+    gain->setUnitsHint(EffectManifestParameter::UnitsHint::Decibel);
+    gain->setRange(-25, 0, 25);
+
+    return pManifest;
+}
+
+CompressorGroupState::CompressorGroupState(
+        const mixxx::EngineParameters& engineParameters)
+        : EffectState(engineParameters),
+          previousStateDB(0),
+          previousAttackParamMs(10),
+          previousAttackCoeff(calculateBallistics(10, engineParameters)),
+          previousReleaseParamMs(150),
+          previousReleaseCoeff(calculateBallistics(150, engineParameters)),
+          previousMakeUpGain(0) {
+}
+
+void CompressorEffect::loadEngineEffectParameters(
+        const QMap<QString, EngineEffectParameterPointer>& parameters) {
+    m_pThreshold = parameters.value("threshold");
+    m_pRatio = parameters.value("ratio");
+    m_pKnee = parameters.value("knee");
+    m_pAttack = parameters.value("attack");
+    m_pRelease = parameters.value("release");
+    m_pGain = parameters.value("gain");
+    m_pAutoMakeUp = parameters.value("automakeup");
+}
+
+void CompressorEffect::processChannel(
+        CompressorGroupState* pState,
+        const CSAMPLE* pInput,
+        CSAMPLE* pOutput,
+        const mixxx::EngineParameters& engineParameters,
+        const EffectEnableState enableState,
+        const GroupFeatureState& groupFeatures) {
+    Q_UNUSED(groupFeatures);
+    Q_UNUSED(enableState);
+
+    SINT numSamples = engineParameters.samplesPerBuffer();
+
+    // Compression
+    applyCompression(pState, engineParameters, pInput, pOutput);
+
+    // Auto make up
+    if (m_pAutoMakeUp->toInt() == static_cast<int>(AutoMakeUp::AutoMakeUpOn)) {
+        applyAutoMakeUp(pState, pOutput, numSamples);
+    }
+
+    // Output gain
+    CSAMPLE gain = static_cast<CSAMPLE>(db2ratio(m_pGain->value()));
+    SampleUtil::applyGain(pOutput, gain, numSamples);
+}
+
+void CompressorEffect::applyAutoMakeUp(CompressorGroupState* pState,
+        CSAMPLE* pOutput,
+        const SINT& numSamples) {
+    double makeUpStateDB = pState->previousMakeUpGain;
+    CSAMPLE maxSample = SampleUtil::maxAbsAmplitude(pOutput, numSamples);
+    if (maxSample > 0) {
+        double maxSampleDB = ratio2db(maxSample);
```

Comment:
I have a warning here: 
```
/home/daniel/workspace/mixxx2/src/effects/backends/builtin/compressoreffect.cpp:165:48:   required from here
/home/daniel/workspace/mixxx2/src/util/math.h:78:21: warning: conversion from ‘double’ to ‘float’ may change value [-Wfloat-conversion]
   78 |     return log10(a) * 20;
      |            ~~~~~~~~~^~~~
```

Reply:
Actually I have no clue, why this warning is still here. I think I have convertion from float to double, but not double to float.

Reply:
ratio2db is a template type. It uses the float implementation because of `CSAMPLE maxSample` 
But log10() is double. So it looks like a static cast to T is missing. 
`return static_cast<T>(log10(a)) * 20;` 
Not you fault, but please fix it. 

The other question is if we need double precision for maxSampleDB, nd can we avoid the DB round trip and instead use the ratio values?

Reply:
Yes, it's a good question. I managed to avoid this DB round trip, so now I don't use ratio2db() and db2ratio() for applying make up.